### PR TITLE
Moving more extensions packages into aspnetcore

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/dotnet/aspnetcore-internal/issues/3707.

We are migrating 
- Microsoft.Extensions.Caching.SqlServer
- Microsoft.Extensions.Caching.StackExchangeRedis
- Microsoft.Extensions.Http.Polly
into aspnetcore from 6.0 and onwards. After this change we no longer own anything in extensions for 6.0 and beyond.

In terms of servicing, we'll need to triage the work and backport fixes to previous releases as necessary for these packages.

### CODE REVIEW

**Comment by JunTaoLuo** [comment]
For reviewers: the only interesting commits are https://github.com/dotnet/aspnetcore/commit/304f206b1d9bca833d6fe0c957178b6a96448913 and https://github.com/dotnet/aspnetcore/commit/a07519084fde70ab95b8f3e39290e22aa04b8d5c (and beyond in case I need to make fixes or address comments)

**Review by wtgodbe — APPROVED** [review]
Everything looks reasonable, LGTM assuming CI works

File: src/Caching/README.md
Comment (dougbu) [inline src/Caching/README.md:1]: Suggest following up to use the template for area-level README files from #27561

File: src/Caching/SqlServer/src/PublicAPI.Shipped.txt
Comment (dougbu) [inline src/Caching/SqlServer/src/PublicAPI.Shipped.txt:1]: Add `#nullable enable`.

The right way to create these files is _not_ in Visual Studio. Instead, copy eng/PublicAPI.empty.txt to PublicAPI.Shipped.txt and PublicAPI.Unshipped.txt in new projects and then fill them up in Visual Studio.

File: src/Caching/SqlServer/src/PublicAPI.Unshipped.txt
Comment (dougbu) [inline src/Caching/SqlServer/src/PublicAPI.Unshipped.txt:1]: These APIs shipped or mostly shipped in 5.0. Move everything that isn't new for 6.0 to the PublicAPI.Shipped.txt file.

Also, add the `#nullable enable` line then recreate this file to get the right annotations.

File: src/Caching/StackExchangeRedis/src/PublicAPI.Shipped.txt
Comment (dougbu) [inline src/Caching/StackExchangeRedis/src/PublicAPI.Shipped.txt:1]: Same comments about the `PublicAPI.*.txt` files as above

File: src/HttpClientFactory/Polly/src/PublicAPI.Shipped.txt
Comment (dougbu) [inline src/HttpClientFactory/Polly/src/PublicAPI.Shipped.txt:1]: Same comments about the `PublicAPI.*.txt` files as above

**Review by dougbu — COMMENTED** [review]
Mainly problems with `PublicAPI.*.txt` files&hellip;

File: src/Caching/README.md
Comment (dougbu) [inline src/Caching/README.md:3]: nit
```suggestion
This area contains distributed cache implementations for Microsoft SQL Server and Redis. See documentation at <https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests#use-polly-based-handlers>.
```

File: src/Caching/README.md
Comment (dougbu) [inline src/Caching/README.md:7]:
```suggestion
These packages are not part of the ASP.NET Core shared framework but are shipped as an individual package. These packages have a third party dependencies on Polly and Polly.Extensions.Http.
```

File: src/Caching/SqlServer/Directory.Build.props
Comment (dougbu) [inline src/Caching/SqlServer/Directory.Build.props:7]: nit: Why are these settings here instead of in the src/Caching/SqlServer/src/Microsoft.Extensions.Caching.SqlServer.csproj file❔

File: src/Caching/StackExchangeRedis/Directory.Build.props
Comment (dougbu) [inline src/Caching/StackExchangeRedis/Directory.Build.props:6]: nit: Again, why is this setting here❔ We have too many `Directory.Build.*` files and they make settings harder to understand.

File: src/HttpClientFactory/Polly/Directory.Build.props
Comment (dougbu) [inline src/HttpClientFactory/Polly/Directory.Build.props:6]: nit: Another Directory.Build.props I'd rather we didn't add.

File: eng/Versions.props
Comment (dougbu) [inline eng/Versions.props:247]: Something about these needs to be included in the root THIRD-PARTY-NOTICES.txt file

**Review by dougbu — CHANGES_REQUESTED** [review]

File: src/Caching/SqlServer/Directory.Build.props
Comment (JunTaoLuo) [reply src/Caching/SqlServer/Directory.Build.props:7]: No particular reason, I just didn't bother changing it when I migrated them from extensions. I'll update them.

**Review by dougbu — APPROVED** [review]

File: THIRD-PARTY-NOTICES.txt
Comment (JunTaoLuo) [inline THIRD-PARTY-NOTICES.txt:377]: FYI, this is probably redundant, we don't include on a per package basis, it's more like per author basis.

Comment (JunTaoLuo) [reply THIRD-PARTY-NOTICES.txt:377]: FYI in offline discussions, it was brought up that we don't need to include it in the 3PN if we are only using nuget package references which is the case for the Polly references. Hence we don't need to change anything in the 3PN.

**Comment by JunTaoLuo** [comment]
There were a few changes to these files after my merge so I'll cherry-pick those over.

**Pull request merged**
